File: src/xml.c
/**
 * @file xml.c
 *
 * @breif Routines and functions for processing a XML file.
 *        This is a very thin layer around libxml2.
 *
 *
 * @ingroup XML
 * @{
 **/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <err.h>
#include <sysexits.h>

#include <libxml/parser.h>
#include <libxml/tree.h>

#include "types.h"
#include "xml.h"

/**
 * Read a xml file and load the information.
 *
 * @param[in]  filename  The xml file name.
 * @param[out] xml       The xml document pointer.
 * @param[out] root      The root node of the xml document.
 * @retval     0         If it was sucessful.
 * @retval     1         If there was an error.
 **/
int
xml_load(const char *filename, void **xml, void **root)
{

	/* Read the file into a document tree */
	*xml = (void *)xmlReadFile(filename, NULL, 0);
	if (*xml == NULL) {
		errx(EX_SOFTWARE, "Failed to parse %s", filename);
	}

	*root = (void *)xmlDocGetRootElement((xmlDocPtr)(*xml));

	return(EXIT_SUCCESS);
}

/**
 * Unload the XML document and clean-up the XML parser.
 *
 * @param[in] xml   The xml document pointer.
 * @retval     0    If it was sucessful.
 * @retval     1    If there was an error.
 **/
int
xml_unload(void **xml)
{
	xmlDocPtr doc = NULL;         /**< XML document tree **/

	doc = (xmlDocPtr)(*xml);

	/* Free the document tree */
	xmlFreeDoc(doc);

	/* Clean up the parser */
	xmlCleanupParser();

	return(EXIT_SUCCESS);
}

/**
 * Get the first occurance of the node.
 *
 * @param[in]  node    The toplevel node pointer to start from.
 * @param[in]  name    The name element to retrieve.
 * @param[out] ele     The first occurance of the element.
 * @retval     0       If it was sucessful.
 * @retval     1       If there was an error.
 **/
int
xml_ele_find(void **node, const char *name, void **ele)
{
	xmlNodePtr cur = NULL;         /**< XML tree root node **/

	cur = (xmlNodePtr)(*node);

	/* Loop through all children finding the first element requested */
	cur = cur->xmlChildrenNode;
	while (cur != NULL) {
		if (xmlStrcmp(cur->name, (const xmlChar *)name) == 0) {
			*ele = (void *)cur;
			break;
		}
		cur = cur->next;
	}
	if (!*ele) {
		return(EXIT_FAILURE);
	}

	return(EXIT_SUCCESS);
}

/**
 * Get the next occurance of the node.
 *
 * This uses xmlNextElementSibling() followed by a check of
 * the name.
 *
 * @param[in]  node    The toplevel node pointer to start from.
 * @param[in]  name    The name element to retrieve.
 * @param[out] ele     The next occurance of the element.
 * @retval     0       If it was sucessful.
 * @retval     1       If there was an error.
 **/
int
xml_ele_next(void **node, const char *name, void **ele)
{
	xmlNodePtr cur = NULL;         /**< XML tree root node **/

	cur = (xmlNodePtr)(*node);

	cur = xmlNextElementSibling(cur);
	/* Loop through all siblings finding the element requested */
	while (cur != NULL) {
		if (xmlStrcmp(cur->name, (const xmlChar *)name) == 0) {
			*ele = (void *)cur;
			break;
		}
		cur = xmlNextElementSibling(cur);
	}
	if (!*ele) {
		return(EXIT_FAILURE);
	}

	return(EXIT_SUCCESS);
}

/**
 * Count the number of elements within the XML node.
 *
 * @param[in]  node    The toplevel node pointer to start from.
 * @param[in]  name    The name element to count.
 * @param[out] n       The number of times the element was found.
 * @retval     0       If it was sucessful.
 * @retval     1       If there was an error.
 **/
int
xml_ele_count(void **node, const char *name, int *n)
{
	xmlNodePtr cur = NULL;         /**< XML tree root node **/

	cur = (xmlNodePtr)(*node);

	*n = 0;

	/* Count the number of elements */
	cur = cur->xmlChildrenNode;
	while (cur != NULL) {
		if (xmlStrcmp(cur->name, (const xmlChar *)name) == 0) {
			++(*n);
		}
		cur = cur->next;
	}

	return(EXIT_SUCCESS);
}

/**
 * Get the contents of a node.
 *
 * @param[in]  node    The toplevel node pointer to start from.
 * @param[out] value   The value of the attribute.
 * @retval     0       If it was sucessful.
 * @retval     1       If there was an error.
 **/
int
xml_ele_contents(void **node, char *(value[STR_LEN]))
{
	int n          = 0;            /**< String length **/
	xmlNodePtr cur = NULL;         /**< XML tree node **/
	xmlChar *tmp   = NULL;         /**< The contents value **/

	cur = (xmlNodePtr)(*node);

	tmp = xmlNodeGetContent(cur);
	if (!tmp) {
		return(EXIT_FAILURE);
	}

	n = strlen((char *)tmp);
	memset(*value, 32, (STR_LEN) * sizeof(char));
	strncpy(*value, (char *)tmp, n * sizeof(char));
	xmlFree(tmp);

	return(EXIT_SUCCESS);
}

/**
 * Get the attribute at the node.
 *
 * @param[in]  node    The toplevel node pointer to start from.
 * @param[in]  name    The name of the attribute to get.
 * @param[out] value   The value of the attribute.
 * @retval     0       If it was sucessful.
 * @retval     1       If there was an error.
 **/
int
xml_ele_att(void **node, const char *name, char *(value[STR_LEN]))
{
	int n          = 0;            /**< String length **/
	xmlNodePtr cur = NULL;         /**< XML tree node **/
	xmlChar *tmp   = NULL;         /**< The attribute value **/

	cur = (xmlNodePtr)(*node);

	tmp = xmlGetProp(cur, (const xmlChar *)name);
	if (!tmp) {
		return(EXIT_FAILURE);
	}

	n = strlen((char *)tmp);
	memset(*value, 32, (STR_LEN) * sizeof(char));
	strncpy(*value, (char *)tmp, n * sizeof(char));
	xmlFree(tmp);

	return(EXIT_SUCCESS);
}


/**
 * @}
 **/

File: src/xml.f90
!>
!! @brief XML functions and subroutines module.
!!
!! @details The XML module provides functions and 
!!          subroutines for accessing the C versions
!!          of the functions.
!
module xml

    use, intrinsic :: iso_c_binding

    implicit none

    private
    public :: xml_load,                                              &
              xml_unload,                                            &
              xml_ele_find,                                          &
              xml_ele_next,                                          &
              xml_ele_contents,                                      &
              xml_ele_count,                                         &
              xml_ele_att

    interface
       integer(c_int32_t)                                             &
       function xml_load(filename, xml, root)                         &
                bind(c, name='xml_load')
           import :: c_int32_t, c_char, c_ptr
           character(kind=c_char), dimension(*) :: filename
           type(c_ptr)                          :: xml
           type(c_ptr)                          :: root
       end function xml_load

       integer(c_int32_t)                                             &
       function xml_unload(xml)                                       &
                bind(c, name='xml_unload')
           import :: c_int32_t, c_ptr
           type(c_ptr)                          :: xml
       end function xml_unload

       integer(c_int32_t)                                             &
       function xml_ele_find(xml, name, ele)                          &
                bind(c, name='xml_ele_find')
           import :: c_int32_t, c_ptr, c_char
           type(c_ptr)                          :: xml
           character(kind=c_char), dimension(*) :: name
           type(c_ptr)                          :: ele
       end function xml_ele_find

       integer(c_int32_t)                                             &
       function xml_ele_next(xml, name, ele)                          &
                bind(c, name='xml_ele_next')
           import :: c_int32_t, c_ptr, c_char
           type(c_ptr)                          :: xml
           character(kind=c_char), dimension(*) :: name
           type(c_ptr)                          :: ele
       end function xml_ele_next

       integer(c_int32_t)                                             &
       function xml_ele_contents(xml, val)                            &
                bind(c, name='xml_ele_contents')
           import :: c_int32_t, c_ptr, c_char
           type(c_ptr)                          :: xml
           type(c_ptr)                          :: val
       end function xml_ele_contents

       integer(c_int32_t)                                             &
       function xml_ele_count(xml, name, n)                           &
                bind(c, name='xml_ele_count')
           import :: c_int32_t, c_ptr, c_char
           type(c_ptr)                          :: xml
           character(kind=c_char), dimension(*) :: name
           integer(c_int32_t)                   :: n
       end function xml_ele_count

       integer(c_int32_t)                                             &
       function xml_ele_att(node, name, val)                          &
                bind(c, name='xml_ele_att')
           import :: c_int32_t, c_ptr, c_char
           type(c_ptr)                          :: node
           character(kind=c_char), dimension(*) :: name
           type(c_ptr)                          :: val
       end function xml_ele_att

   end interface

end module xml
